### PR TITLE
target: Add support for LPC40xx

### PR DESCRIPTION
**Pull request by zyp**
## Detailed description

Add support for LPC40xx. This is mainly a copy of LPC17xx support with a different probing path and different IAP PARTIDs.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

### CODE REVIEW

File: src/target/lpc40xx.c [hunk 1]
```diff
@@ -0,0 +1,273 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2012 Gareth McMullin <gareth@blacksphere.co.nz>
```
Comment (dragonmux) [inline src/target/lpc40xx.c:4]: You may wish to update this to credit yourself and have the correct copyright year

File: src/target/lpc40xx.c [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+	flash->base_sector = base_sector;
+	flash->f.write = lpc_flash_write_magic_vect;
+	flash->iap_entry = IAP_ENTRYPOINT;
+	flash->iap_ram = IAP_RAM_BASE;
+	flash->iap_msp = IAP_RAM_BASE + MIN_RAM_SIZE - RAM_USAGE_FOR_IAP_ROUTINES;
+}
+
+bool lpc40xx_probe(target_s *target)
+{
+	if ((target->cpuid & CPUID_PARTNO_MASK) == CORTEX_M4) {
```
Comment (dragonmux) [inline src/target/lpc40xx.c:89]: Would this make more sense as a check for `!= CORTEX_M4` and a fast `return false;` so the rest of the code can be less indented?

**Review by dragonmux — CHANGES_REQUESTED** [review]
This looks good - there are a couple of things that stood out in review but once those have been taken care of, we're happy to merge this

File: src/target/lpc40xx.c (hunk 2 above)
Comment (zyp) [reply src/target/lpc40xx.c]: Fair point. I also fixed the lpc17xx.c that this was copied from.

**Review by dragonmux — APPROVED** [review]
LGTM, we'll merge this when the builds complete.

Thank you for the contribution!

**Pull request merged**